projects/evm-compatible-bridge/federator.md

```diff
@@ -0,0 +1,245 @@
+
+The federator service will keep track of events on both sides of the bridge and vote on the actions to be taken.
+This will be done by watching events on the federation contract on the EVM chain and polling events from the coordinator service on Hathor.
+
+Since Hathor does not currently have a smart contract implementation, we will use a MultiSig address to check that a certain number of federators have consented to any transaction.
+To avoid separate instances of the MultiSig wallet to propose conflicting transactions, we will use a coordinator service to choose the utxos and collect signatures from the federators.
+
+Each federator will have 2 loops, one to check events on the EVM compatible chain (i.e. EVM loop) and another to check events on the coordinator service (i.e. polling loop).
+The EVM loop will get the events from the chain, this loop will have events of transaction requests from the EVM compatible chain to Hathor and confirmations on transactions.
+The polling loop will check the coordinator service polling api and vote to perform the actions as requested.
+
+## Federator service
+
+### EVM loop
+
+We need to get events from the bridge contract, wait a pre-configured number of blocks (to increase the confirmation level of the transaction) and send the transaction on Hathor as requested.
+There should be 2 types of events:
+
+- A request to cross the token from the EVM chain to Hathor
+  - If the token is native to the EVM chain, an equivalent Hathor token should already exist and the MultiSig wallet should own the authority utxos to it.
+  - If the token is native to the Hathor chain, the equivalent token is already burned and we need to unlock the tokens on Hathor.
+  - The loop will have to use the coordinator service to send the amount of tokens requested from the MultiSig wallet to the destination address.
+- Confirmation that an operation is complete
+  - This way we can clean any state pertaining to this crossing of tokens and mark it as complete.
+  - If the crossing was originated in Hathor, we need to send a request to the coordinator service so it knows the crossing is over.
+
+### Polling loop
+
+We will pool for events, which are requests for voting, and vote on them.
+If the event requires a vote on the smart contract, we will send a transaction to the EVM chain with the vote.
+If the event requires a vote on the MultiSig wallet, we will send a message to the coordinator service with a signature.
+
+The federator will be responsible to check the data on the event, it will have its own headless wallet to check the information from the Hathor network and it can read the data from the EVM chain.
+
+## Coordinator service
+
+The coordinator service will serve the function of the smart contract in the Hathor side.
+It will listen for transactions on the MultiSig wallet, keep a comprehensive persistent database so we do not process the same transaction twice and will gather signatures from the federation.
```

Comment:
I guess this answers my question above. It doesn't seem like a good solution, since players would have to trust the coordinator for information. The federators should always be able to verify the information independently.

Reply:
Below there is a passage that the coordinator will check the database and if not present it will check the contract, since crossing confirmations will be saved on the contract (indexed by the transaction id). We can remove the check from the database and read directly from the contract everytime, what do you think?

Reply:
But this still has the problem that the multisig wallets will have to trust the coordinator, right?

Reply:
I'll change to only check against the chain.
The power of veto on transactions being used maliciously (saying the transaction was made when it was not) may not have dire consequences but i agree that the simpler approach (always checking the chain) is better.

While not used by the federation I think we should keep the check that the request was fulfilled on the database, it would make it easier to show the admin or create a summary of the operations the bridge has made without the need to check the chain on all operations.

Reply:
I agree.